.github/workflows/cmake_build.yml

```
@@ -63,14 +77,34 @@ jobs:
     - name: Environment
       run: |
         source "scripts/ci/environment.sh"
-    
+
     - name: Install Apple codesigning certificates
       uses: apple-actions/import-codesign-certs@v1
       if: startswith( matrix.config.os, 'macos' ) && github.event_name == 'push' && github.repository_owner == 'audacity'
       with: 
         p12-file-base64: ${{ secrets.APPLE_CERTIFICATE }}
         p12-password: ${{ secrets.APPLE_CERTIFICATE_PASSWORD }}
 
+    # required for CMake to find Ninja
+    - name: "[Windows] Set up MSVC Developer Command Prompt"
+      if: runner.os == 'Windows'
+      uses: seanmiddleditch/gha-setup-vsdevenv@v3
+      with:
+        arch: ${{ matrix.config.arch }}
+
+    - name: "[Windows] Install sccache"
+      if: runner.os == 'Windows'
+      run: choco install sccache
+
+    - name: "Set up compiler cache"
+      uses: actions/cache@v2
+      with:
+        path: ${{ matrix.config.compiler_cache_path }}
+        key: ${{ matrix.config.os }}-${{ matrix.config.arch }}-${{ github.head_ref }}-${{ github.run_number }}
+        restore-keys: |
+          ${{ matrix.config.os }}-${{ matrix.config.arch }}-${{ github.head_ref }}
+          ${{ matrix.config.os }}-${{ matrix.config.arch }}
```

This seems to indicate that all new commits will be built using cache from `${{ matrix.config.os }}-${{ matrix.config.arch }}`. But where is this cache ID populated?

---

I don't understand your question. The cache is populated by building.

---

My question is: if I make an absolutely new commit, will it use cached data? Presumably yes, and the only cache ID that does not depend on commit ID is `${{ matrix.config.os }}-${{ matrix.config.arch }}`, so every build of a new commit starts with initial cache data loaded from there *only*, right? But I don't see where `${{ matrix.config.os }}-${{ matrix.config.arch }}` is *written* to, that's my question.

---

https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows

---

Thanks for asking me to RTFM but I've done that before. That manual says to use something like `${{ runner.os }}-build-${{ env.cache-name }}-${{ hashFiles('**/package-lock.json') }}` which makes sense because multiple commits with the same `package-lock.json` file will reuse the same cache ID. This PR uses `${{ github.head_ref }}` so the same logic does not apply.

---

Do I understand it correctly that ccache and sccache check that the file has not been changed since it was built, including its dependencies like includes? If yes, we could probably use something like this instead:
```suggestion
        key: ${{ matrix.config.os }}-${{ matrix.config.arch }}-${{ hashFiles('**/CMakeLists.txt') }}
        restore-keys: |
          ${{ matrix.config.os }}-${{ matrix.config.arch }}
```

---

> When key doesn't match an existing cache, it's called a cache miss, and a new cache is created if the job completes successfully. When a cache miss occurs, the action searches for alternate keys called restore-keys.
> If you provide restore-keys, the cache action sequentially searches for any caches that match the list of restore-keys.
> When there is an exact match, the action restores the files in the cache to the path directory.
> If there are no exact matches, the action searches for partial matches of the restore keys. When the action finds a partial match, the most recent cache is restored to the path directory.

---

> check that the file has not been changed

What file? I don't know what you're talking about. There is no file involved in this cache key.

---

> What file? I don't know what you're talking about. There is no file involved in this cache key.

I'm asking about how ccache works, because I'm not familiar with it.

---

> When key doesn't match an existing cache, it's called a cache miss, and a new cache is created if the job completes successfully. When a cache miss occurs, the action searches for alternate keys called restore-keys.

Sorry, but did you actually try to read what I said? My questions were about where those restore-keys caches were populated. Then you linked me to a manual to (probably) prove that it's fine and I replied that the manual doesn't do anything remotely similar to what you did and suggested using `hashFiles` like how the manual did, like how I have done before and like what seems like the best solution to me.

---

> I'm asking about how ccache works, because I'm not familiar with it.

Then please familiarize yourself with it instead of suggesting nonsensical changes to functioning code.

---

Can you avoid calling me out please and answer my question about where the restore-keys are populated from? That is certainly unrelated to ccache.

Also my question about ccache was (supposedly) simple enough to answer in two or three words if you are yourself familiar with it. My question was whether ccache recompiles source file when its included file is changed. I think the answer is yes because otherwise no one would use ccache, but I just wanted to make sure.

Suppose I have the following file structure:
**main.c:**
```
#include "header.h"
```
**header.h:**
```
// hello
```

If I change header.h only, will ccache recompile main.c?

---

> the manual doesn't do anything remotely similar to what you did and suggested using hashFiles

ccache and sccache take care of tracking whether the source code changes. `hashFiles` is for cases where the contents of the cache depend on one specific file, for example a Cargo.lock file for caching the dependencies of a Rust project.

---

I do expect that if you review code you have a basic familiarity with the topic of the code changes or read the documentation for whatever software is involved to learn about it. If the documentation is missing or unclear, then ask about that, but it comes across as rude to ask me to teach you how ccache works when the topic of this PR is using ccache.

---

> ccache and sccache take care of tracking whether the source code changes. hashFiles is for cases where the contents of the cache depend on one specific file, for example a Cargo.lock file for caching the dependencies of a Rust project.

Right -- this way you cache exactly what depends on the hashed file and don't cache anything that depends on something else. However, this is not going to work for our use case.

Consider this. Say I have something in `${{ matrix.config.os }}-${{ matrix.config.arch }}` (or maybe `os-arch-headref`, for that matter) at the moment. Then I commit some drastic change that requires rebuilding lots of files and trigger CI. The `os-arch-headref-runnumber` thingy doesn't exist yet, so GitHub loads cache from `os-arch` or `os-arch-headref`, whichever exists. Lots of files were changed so the build is slow, nothing unexpected yet. Then when the build is finished the cache for `osarch-headref-runnumber`--and only it--is saved. Then pretend I add yet another very simple commit, like changing just a single file. `runnumber` was definitely changed so `os-arch-headref-runnumber` doesn't exist yet, so GitHub resorts to `os-arch` or `os-arch-headref`, again. Those cache keys were populated before that drastic commit, so the cached files are again very different from the files in this small commit, so the build is slow again. Then I may wait to make yet another simple commit, and again GitHub will load cache from `os-arch` or `os-arch-headref`, which were populated in the far past and so everything has to be rebuilt, and so on.

The problem is that `os-arch` and `os-arch-headref` are populated *once*, at some point of time, and never updated by CI afterwards, so they will be slowly (or not) getting more and more different from new commits, and the build will become slower and slower.

---

You're misunderstanding how GitHub Actions works. There is no cache with a key of `os-arch`. The cache key is specified by:
```
        key: ${{ matrix.config.os }}-${{ matrix.config.arch }}-${{ github.head_ref }}-${{ github.run_number }}
```

The last run on this branch had a cache hit:
```
Cache restored from key: ubuntu-18.04-x86_64-windows_sccache-163
```
and at the end it uploaded an updated cache:
```
Cache saved with key: ubuntu-18.04-x86_64-windows_sccache-165
```

>  the build will become slower and slower.

This is not correct, as demonstrated by the builds on this PR as well as the same setup on Mixxx.

---

> You're misunderstanding how GitHub Actions works.

Fuck me. That's not what I got wrong but I just found what confused me so badly. I'm very sorry. I thought `restore-keys` matched not a prefix but a whole string by default, and only resorted to prefix if the key ended with a dash. I think the GitHub examples make this a bit more intuitive, so what about adding dashes so that no one else gets confused (no one would end a cache key with a dash so this would be thought-provoking)?
```yml
restore-keys: |
    ${{ matrix.config.os }}-${{ matrix.config.arch }}-${{ github.head_ref }}-
    ${{ matrix.config.os }}-${{ matrix.config.arch }}-
```

I think this can be marked as resolved now.

As for ccache... I've got some intuition about how things work, and I've worked with stuff like ccache some time ago, so I half-remembered half-guessed how it would handle include's. Sometimes I'm almost certain I've grasped some topic but I'm like 95% sure, not 100% sure, so I make up the simplest possible example and ask a person who knows the topic thoroughly whether I'm right--that's much better than reading the documentation because the docs may be confusing or erroneous, but a professional has probably stumbled upon a similar problem already and knows the answer for sure. That's what I tried here: I guessed that ccache handled the include dependencies correctly but I wasn't 100% sure so I made a very simple example and asked you whether it works as I expected. You didn't answer with 'yes' or 'no' or 'idk try yourself' but just started name calling.

I see we've got some strained relations here, although I have no idea why it's mostly about me and you. I admit I was wrong on some points. But I am willing to learn from those who know something better than me, especially if I'm going to work on this codebase for a long while alongside you, and it shocks me how you are apparently not trying to comprehend my questions--if something seems wrong to you you repeatedly say the same thing over and over instead of trying to find what's wrong or saying like 'i have no time atm sorry please figure it out yourself'. It feels awkward that you reject my simple questions with 'RTFM' even though you would supposedly be able to answer it if you spent one more minute reading what I said.

---

I don't know what provoked this, but I want this concern resolved, so please let's figure that out. Either you treat me as a junior and ask me to treat yourself as a senior who has no time for anyone's question and I read all the docs three times before asking the simplest question--if that's easier for you. Or you think I'm hopeless and you don't want me reading or commenting on or reviewing your PRs at all. Or you start behaving a bit more civilly with me and try to comprehend and correct what I say, even when I'm somewhat wrong. Honestly this is starting to get on my nerves and I don't want an open-source project to cause depression and whatnot, so if we have a conflict again while fixing this conflict I'm quitting for good. Don't or do take that personally, whatever is easier for you.

---

Sorry for being impatient; I'm frustrated just trying to get this code to build and figuring out some way out of the mess Audacity created.

---

What annoyed me about your comments on this code was implying that this code had to be changed because it wasn't working, but it was working as I mentioned in comments above and you could confirm yourself by checking the build times.